torch_xla/distributed/fsdp/utils.py

```diff
@@ -122,7 +122,9 @@ def _xla_patched_nn_linear_forward(m, input):
   return XLAPatchedLinear.apply(input, m.weight, m.bias)
 
 
-def apply_xla_patch_to_nn_linear(module):
+def apply_xla_patch_to_nn_linear(module,
```

Comment:
will this function always be called? or only when fsdp being used.

Reply:
It's called in the FSDP wrapper. For GSPMD, we don't have a wrapper so we have to call it explicitly.

Reply:
@alanwaketan did it, using einsum instead of matmul, improve the benchmark training result? I recall that it didn't actually have an impact (or lowered MFU in @jonb377 's result?)? If so, we should make a comment explaining that this may reduce the performance, too. And +1 to not enabling by default...

Reply:
The lower MFU was from user-mode FSDP, I don't have a head-to-head benchmark on SPMD. @alanwaketan had some data to suggest it improves performance for SPMD.

Reply:
This is the solution in our end to stop the reshape on nn.Linear's matmul.